### PR TITLE
[TF] Update TF v2.16.1

### PR DESCRIPTION
**Pull request by smuzaffar**
Update tensorflow version to 2.16.1

### CODE REVIEW

**Comment by smuzaffar** [comment]
please test for CMSSW_14_1_TF_X

**Comment by smuzaffar** [comment]
please test for CMSSW_14_1_TF_X

**Comment by smuzaffar** [comment]
please test for CMSSW_14_1_TF_X

**Comment by smuzaffar** [comment]
please test for CMSSW_14_1_TF_X

**Comment by iarspider** [comment]
please test for CMSSW_14_1_TF_X

**Comment by iarspider** [comment]
please test for CMSSW_14_1_TF_X/el8_amd64_gcc12

**Comment by iarspider** [comment]
please test for CMSSW_14_1_TF_X/el8_aarch64_gcc12

**Comment by smuzaffar** [comment]
@iarspider , try updating scipy (https://github.com/scipy/scipy/pull/18242)

**Comment by iarspider** [comment]
please test for CMSSW_14_1_TF_X/el8_amd64_gcc12

**Comment by iarspider** [comment]
please test for CMSSW_14_1_TF_X/el8_amd64_gcc12

**Comment by iarspider** [comment]
please test for CMSSW_14_1_TF_X/el8_amd64_gcc12
blosc2 2.5.1 was the last version to support Python 3.9

**Comment by iarspider** [comment]
please test for CMSSW_14_1_TF_X/el8_amd64_gcc12

**Comment by iarspider** [comment]
~~Testing locally with new cudnn, _maybe_ this will help.~~

**Comment by iarspider** [comment]
please test for CMSSW_14_1_TF_X/el8_amd64_gcc12

**Comment by iarspider** [comment]
please test

**Comment by iarspider** [comment]
please test for CMSSW_14_1_TF_X/el8_amd64_gcc12

**Comment by iarspider** [comment]
Tests got stuck, retried.

**Comment by iarspider** [comment]
please test for CMSSW_14_1_TF_X/el8_amd64_gcc12

**Comment by iarspider** [comment]
please test with https://github.com/cms-sw/cmsdist/pull/9360, https://github.com/cms-sw/cmssw/pull/45701

**Comment by smuzaffar** [comment]
please test with https://github.com/cms-sw/cmssw/pull/45701 for CMSSW_14_1_TF_X/el8_amd64_gcc12

**Comment by smuzaffar** [comment]
please test with https://github.com/cms-sw/cmssw/pull/45701 for CMSSW_14_1_TF_X/el8_aarch_gcc12

**Comment by smuzaffar** [comment]
please test with https://github.com/cms-sw/cmssw/pull/45701 for CMSSW_14_1_TF_X/el8_aarch64_gcc12

**Comment by smuzaffar** [comment]
@iarspider , can you please update `keras` version 3? I think unit tests are failing are tf 2.16 moved to keras 3 ( see https://discuss.huggingface.co/t/pretrain-model-not-accepting-optimizer/76209/6 )

**Comment by iarspider** [comment]
please test with https://github.com/cms-sw/cmssw/pull/45701 for CMSSW_14_1_TF_X/el8_amd64_gcc12

**Comment by iarspider** [comment]
please test with https://github.com/cms-sw/cmssw/pull/45701 for CMSSW_14_1_TF_X/el8_amd64_gcc12

File: tensorflow-xla-runtime.spec [hunk 1]
```diff
@@ -25,6 +25,8 @@ CXXFLAGS="-fPIC -Wl,-z,defs %{arch_build_flags} ${CMS_EIGEN_CXX_FLAGS} %{selecte
 pushd tensorflow/xla_aot_runtime_src
   # remove unnecessary implementations that use symbols that are not even existing
   find . -type f -path '*/service/cpu/runtime_fork_join.cc' | xargs rm -f
+  find . -type f -path '*/service/cpu/runtime_fft.cc' | xargs rm -f
+  find . -type f -path '*/service/cpu/runtime_single_threaded_fft.cc' | xargs rm -f
```
Comment (smuzaffar) [inline tensorflow-xla-runtime.spec]: @iarspider , the fft symbols are provided by tensorflow/thrid_part/ducc. Can you please add 
```
bazel $BAZEL_OPTS @ducc//:fft
bazel $BAZEL_OPTS @ducc//:fft_wrapper
```
in tensorflow-sources. This should generate 
```
> ls bazel-bin/external/ducc/*.so
bazel-bin/external/ducc/libfft.so  bazel-bin/external/ducc/libfft_wrapper.so
```
and we can then install these under tensorflow package and update xla_aot_runtime_src  to link against these libs

**Comment by iarspider** [comment]
please test with https://github.com/cms-sw/cmssw/pull/45701 for CMSSW_14_1_TF_X/el8_amd64_gcc12

File: tensorflow-sources.file [hunk 2]
```diff
@@ -155,6 +155,8 @@ bazel $BAZEL_OPTS //tensorflow/compiler/tf2xla:xla_compiled_cpu_function
 bazel $BAZEL_OPTS //tensorflow/core/profiler
 bazel $BAZEL_OPTS //tensorflow:install_headers
 bazel $BAZEL_OPTS //tensorflow/compiler/tf2xla:tf2xla_supported_ops
+bazel $BAZEL_OPTS @ducc//:fft
+bazel $BAZEL_OPTS @ducc//:fft_wrapper
```
Comment (smuzaffar) [inline tensorflow-sources.file]: humm, may be move these on top ( as first thing to build) ?

**Comment by iarspider** [comment]
please test with https://github.com/cms-sw/cmssw/pull/45701 for CMSSW_14_1_TF_X/el8_amd64_gcc12

**Comment by iarspider** [comment]
please test for CMSSW_14_1_TF_X/el8_amd64_gcc12

**Comment by iarspider** [comment]
@smuzaffar is this the right way to fix compilation failure?

**Comment by smuzaffar** [comment]
@iarspider , I do not think you need these fft libs in cmssw. Try linking both `fft and fft_wrapper` to `libtf_xla_runtime.so` (and I would suggest to use CMAKE_LINKER_FLAG=`-L/tf/path/lib -lfft -lfft_wrapper`).

**Comment by iarspider** [comment]
please test for CMSSW_14_1_TF_X/el8_amd64_gcc12

File: tensorflow-xla-runtime.spec [hunk 3]
```diff
@@ -30,7 +30,7 @@ pushd tensorflow/xla_aot_runtime_src
     -DCMAKE_CXX_FLAGS="${CXXFLAGS}" \
     -DCMAKE_CXX_STANDARD=%{cms_cxx_standard} \
     -DCMAKE_PREFIX_PATH=${ABSEIL_CPP_ROOT} \
-    -DLIBFFT_WRAPPER=${TENSORFLOW_ROOT}/lib/libfft_wrapper.so \
+    -DCMAKE_LINKER_FLAG=-L${TENSORFLOW_ROOT}/lib -lfft -lfft_wrapper
```
Comment (smuzaffar) [inline tensorflow-xla-runtime.spec]: you might want to use double quotes here `-DCMAKE_LINKER_FLAG="-L${TENSORFLOW_ROOT}/lib -lfft -lfft_wrapper"`

Comment (smuzaffar) [reply tensorflow-xla-runtime.spec]: and missing `\` at the end

Comment (smuzaffar) [reply tensorflow-xla-runtime.spec]: also double check the correct cmake flag name for linker flags

**Comment by iarspider** [comment]
please test for CMSSW_14_1_TF_X/el8_amd64_gcc12

**Comment by iarspider** [comment]
please test for CMSSW_14_1_TF_X/el8_amd64_gcc12

**Comment by iarspider** [comment]
please test for CMSSW_14_2_TF_X/el8_amd64_gcc12

**Comment by iarspider** [comment]
please test for CMSSW_14_2_TF_X/el8_amd64_gcc12

**Comment by iarspider** [comment]
* The static libraries don't have fixed location, so `find` is the only solution
* I couldn't find a way to package these libraries inside `whl` without repacking it, thus build-time dependency on tf-sources. Let me know if you want me to repack wheel to add them.

**Comment by iarspider** [comment]
please test for CMSSW_14_2_TF_X/el8_amd64_gcc12

**Comment by iarspider** [comment]
please abort

**Comment by iarspider** [comment]
please test for CMSSW_14_2_TF_X/el8_amd64_gcc12

**Comment by iarspider** [comment]
Test results (before two cosmetic changes): https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-971011/41396/summary.html

Only one unit test failed, but I don't think it's related to this PR.

**Comment by smuzaffar** [comment]
gpu unit tests are still failing (https://github.com/cms-sw/cmsdist/pull/9388#issuecomment-2330681465)

**Comment by iarspider** [comment]
please test for slc7_amd64_gcc12

**Comment by iarspider** [comment]
@smuzaffar the tests triggered correctly (for 14_2_X/slc7, not for nonexistent 14_2_TF_X/slc7), but looks like it's stuck (no progress for 12h ?)

**Comment by iarspider** [comment]
please test for slc7_amd64_gcc12

**Comment by iarspider** [comment]
please test for slc7_amd64_gcc12

**Comment by iarspider** [comment]
please test for CMSSW_14_2_TF_X/el8_amd64_gcc12

**Comment by iarspider** [comment]
please test for slc7_amd64_gcc12

**Comment by smuzaffar** [comment]
@iarspider , is this ready to go in for TF IBs?

**Comment by iarspider** [comment]
@smuzaffar yes

**Comment by smuzaffar** [comment]
+externals

lets get this in TF_X IBs

**Pull request merged**